lib/csv_export.rb

```diff
@@ -0,0 +1,72 @@
+module CsvExport
+  require 'csv'
+  require 'zip'
+
+  PROTECTED_COLS = %w(id encrypted_password reset_password_token
+                      reset_password_sent_at)
+
+  # generates a csv from the given model data
+  def generate_csv(data, columns = [])
+    columns = data.first.attributes.keys if columns.empty?
+
+    PROTECTED_COLS.each { |col| columns.delete(col) }
+
+    CSV.generate(headers: true) do |csv|
+      csv << columns
+
+      data.each do |o|
+        csv << columns.map do |attr|
+          s = o.send(attr)
+          s.is_a?(ActiveRecord::Base) ? s.name : s
+        end
+      end
+    end
+  end
+
+  # generates a zip file containing multiple CSVs
+  # expects tables to be an array of arrays with the following format:
+  # [[objects, columns], ...]
+  # where columns is optional; defaults to all columns except id
+  def generate_zip(tables)
+    # create the CSVs
+    csvs = tables.map { |model| generate_csv(*model) }
+
+    Zip::OutputStream.write_buffer do |stream|
+      csvs.each_with_index do |csv, i|
+        model_name = tables[i].first.first.class.name
+        stream.put_next_entry "#{model_name}_#{Time.zone.now.to_s(:number)}.csv"
+        stream.write csv
+      end
+    end.string
+  end
+
+  # downloads a csv of the given model table
+  # expects table to be an array with the following format:
+  # [objects, columns]
+  # where columns is optional; defaults to all columns except id
+  def download_csv(data, columns, filename)
+    send_data(generate_csv(data, columns),
```

Comment:
hmm, this isn't so much a problem but I was a bit confused about where `send_data` came from since it's not vanilla Ruby. There's nowhere in this module that makes the dependency on ActionController clear; I'm not sure what best practice is, but I feel like at least a comment before these two methods would be useful. We might even want to include an explicit `require` call at the top to make it even clearer. Let me know what you think.